### PR TITLE
[1.x] Subscription Modifiers

### PR DESCRIPTION
**Pull request by driesvints**
Picks up the work from https://github.com/laravel/cashier-paddle/pull/90 by @yoeriboven. I've sent this in as a PR with a branch on this repo so we can verify that the tests pass.

Closes https://github.com/laravel/cashier-paddle/issues/71

### CODE REVIEW

**Comment by driesvints** [comment]
@yoeriboven About the sync command: I'd maybe indeed add an extra `paddle_subscription_id` to the database table. That should simplify things.

However, I want to first check in if it does makes sense that we're keeping the records in a DB table ourselves. There's an easy endpoint we can use for that where we can filter by subscription ID. Maybe it's just better to implement that? It would simplify the code of this PR a lot and we can get rid of the extra table that way.

File: src/Subscription.php [hunk 1]
```diff
@@ -664,14 +664,14 @@ public function modifiers()
...
     /**
-     * Get a modifier instance by name.
+     * Get a modifier instance by ID.
      *
      * @param  int  $id
      * @return \Laravel\Paddle\Modifier|null
      */
     public function modifier($id)
     {
-        return $this->modifiers->find($id);
```
Comment (yoeriboven) [inline src/Subscription.php]: Won't this create an N+1 problem like [here previously](https://github.com/laravel/cashier-paddle/pull/83)?

File: src/Subscription.php [hunk 2]
```diff
@@ -664,14 +664,14 @@ public function modifiers()
...
+        return $this->modifiers()->find($id);
```
Comment (yoeriboven) [inline src/Subscription.php:674]: Won't this create an N+1 issue like [here](https://github.com/laravel/cashier-paddle/pull/83)?

**Comment by yoeriboven** [comment]
Hadn't thought about that yet, but it seems like the right decision. The only negative I can think of right now is that if you show the modifiers on a page that is often loaded, the modifiers will be downloaded often. That shouldn't be a problem though, since modifiers aren't something you show often.

I guess this could be implemented [like the transactions were](https://github.com/laravel/cashier-paddle/pull/40/files#diff-70aeecbc6b54b94a5804ad16455bc9ed3be130e5c6984e743976088bbbd36d53L669-L685) before they were refactored to receipts.

If you decide it's the right choice and want me to do the refactor: let me know. Happy to help. :)

File: src/Subscription.php (hunk 2 above)
Comment (driesvints) [reply src/Subscription.php]: Ah good catch. I keep forgetting about that one 😑

**Comment by driesvints** [comment]
@yoeriboven yes that seems like a good idea, go for it. In any case if performance is a problem it can always be solved with caching in user land.

**Comment by driesvints** [comment]
Hey @yoeriboven, did you have time to do these changes?

**Comment by GrahamCampbell** [comment]
I think some upgrading documentation will be in order too, since people will need to do a sync after upgrading, so new incoming hooks don't make a mess of the database?

**Comment by yoeriboven** [comment]
Whoops. Worked on this but ran into some issues and then completely forgot to pick it up again.

Will send something in this week.

**Comment by driesvints** [comment]
> I think some upgrading documentation will be in order too, since people will need to do a sync after upgrading?

No, as we discussed above we're going to update this PR so there's no upgrade steps or even database table needed.

File: src/Subscription.php [hunk 3]
```diff
@@ -671,7 +677,9 @@ public function modifiers()
      */
     public function modifier($id)
     {
-        return $this->modifiers->find($id);
+        return $this->modifiers()->first(function (Modifier $modifier) use ($id) {
+            return $modifier->id() == $id;
```
Comment (GrahamCampbell) [inline src/Subscription.php:681]: I think we probably want a strict comparison here?

File: tests/Feature/ModifiersTest.php [hunk 4]
```diff
@@ -35,21 +94,23 @@ public function test_subscriptions_can_create_modifiers()
...
+        $this->assertEquals($modifier->id(), 6789);
+        $this->assertEquals($modifier->subscription()->paddle_id, $subscription->paddle_id);
+        $this->assertEquals($modifier->amount(), 15.00);
+        $this->assertEquals($modifier->currency(), 'EUR');
+        $this->assertEquals($modifier->description(), 'Our test description');
+        $this->assertEquals($modifier->recurring(), false);
+
         Http::assertSent(function ($request) {
-            return $request['subscription_id'] == $_SERVER['PADDLE_TEST_SUBSCRIPTION'] &&
+            if ($request->url() == 'https://vendors.paddle.com/api/2.0/subscription/modifiers/create') {
```
Comment (GrahamCampbell) [inline tests/Feature/ModifiersTest.php:105]: Same

File: tests/Feature/ModifiersTest.php [hunk 5]
```diff
@@ -70,25 +131,18 @@ public function test_a_modifier_can_delete_itself()
             'quantity' => 1,
         ]);
 
-        $modifier = $subscription->modifiers()->create([
-            'paddle_id' => 6789,
-            'amount' => 15.00,
-            'description' => 'Our test description',
-            'recurring' => false,
+        $modifier = new Modifier($subscription, [
+            'modifier_id'       => 6789,
```
Comment (GrahamCampbell) [inline tests/Feature/ModifiersTest.php:135]: Indentation is too far over.

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: tests/Feature/ModifiersTest.php (hunk 5 above)
Comment (driesvints) [reply tests/Feature/ModifiersTest.php]: Resolved

File: src/Subscription.php (hunk 3 above)
Comment (driesvints) [reply src/Subscription.php]: Done

File: tests/Feature/ModifiersTest.php (hunk 4 above)
Comment (driesvints) [reply tests/Feature/ModifiersTest.php]: Done

**Comment by driesvints** [comment]
@yoeriboven I've updated the PR a bit so the "amount" and "currency" methods on the Modifier object are in line with the Price object. I've also made "amount" a required argument for the `newModifier` method as it cannot be optional.

File: src/Subscription.php [hunk 6]
```diff
@@ -643,6 +643,46 @@ public function nextPayment()
...
+    }
+
+    /**
+     * Get all of the modifiers for this subscription.
+     *
+     * @return \Illuminate\Support\Collection
+     */
+    public function modifiers()
+    {
+        $result = Cashier::post('/subscription/modifiers', array_merge([
```
Comment (GrahamCampbell) [inline src/Subscription.php:664]: Does Paddle paginate that endpoint?

File: src/Modifier.php [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    {
+        return $this->modifier['amount'];
+    }
+
+    /**
+     * Get the currency.
+     *
+     * @return \Money\Currency
+     */
+    public function currency(): Currency
```
[the remark this answers is not in the document]
Comment (driesvints) [reply src/Modifier.php:81]: Yes.

File: src/Subscription.php (hunk 6 above)
Comment (driesvints) [reply src/Subscription.php:664]: Nope: https://developer.paddle.com/api-reference/subscription-api/modifiers/listmodifiers

**Comment by driesvints** [comment]
Thanks for this @yoeriboven!

**Pull request merged**